### ISSUE TITLE
Solve time zone problems with clients

### ISSUE THREAD

**Issue by LukasReschke** [nextcloud/server#2700]
On c.nc.c the fact that some people here are (travelling) in a different timezone right now causes a folder to mark files as "in 6 hours".

![2016-12-15_16-49-13](https://cloud.githubusercontent.com/assets/878997/21230830/794ec734-c2e6-11e6-96d5-2953645c06e2.png)

This breaks now the desktop client for people in another timezone such as @nickvergessen. With the message: "Dieser Sync-Vorgang des Synchronisationsordners 'Nextcloud/Company' würde Dateien auf ein früheres Änderungsdatum zurücksetzen."

In my opinion we should either:

1. Don't trust the clients on time at all. Why do we even need that? Others also just take the time when something was uploaded or so.
2. If the time is in the future change it to the current time in the server on upload.

I'd honestly prefer No. 1 myself though. But probably I miss something totally as well 🙈  – So any input is welcome and we should think a lot about the consequences 😉 

This is also kinda related to the issues that we encountered with summer- and winter-time change switches.

---

cc @karlitschek @rullzer @nickvergessen Input welcome.

**Comment by rullzer** [comment]
1. We can't use the upload time. lets say I added files when I was on the train without connection. My system shows the time X, but upload is X+1. Then my client will sync down the file again (since the time is newer) and all of sudden it shows that the file is updated at X+1 on my machine. While nobody ever touched it.

2. So same goes actually for time in the future. My machine is in a different time zone. So I upload at local time X. Now the server changes this to X-5. My machine syncs. And sees a time difference. What to do. Better upload again.


Long story short. Timing and syncing is hard.

**Comment by LukasReschke** [comment]
Thought so I probably miss something - thanks for clarification! :)

Mhm. So I suppose at some point we need to invest some more thinking in how to avoid cases like that :-/

**Comment by nickvergessen** [comment]
Well if a time is sent with the sync, it should contain the timezone? and then we can calculate it back to UTC?

**Comment by schiessle** [comment]
How about always store, sync and compare the timestamp and only the user interface calculates it back to the timezone of the client to display the date?

**Comment by rullzer** [comment]
You assume that all machines have timezones correctly set. And users never just adjust the clock :P

**Comment by jancborchardt** [comment]
Also while we’re at it, just saying that we should never ever display something stupid in the UI like a modification time of »in 6 hours«. Time-travel compliancy just looks ridiculous, and even though »just now« is technically wrong, it makes more sense.

cc @MorrisJobke since you defended it so strongly. ;)

**Comment by MorrisJobke** [comment]
> cc @MorrisJobke since you defended it so strongly. ;)

Because it is an underlying issue. We need to fix the cause not the symptoms ;)

**Comment by pquan** [comment]
The timezone and time-difference between clients must be fixed because it can leads to huge corruption specially if you have folders shared between different persons, same timezone or not.

Say Person A in in UTC+1 and person B is in UTC+2. Person A modified the "sales.xls" and saved to the shared folder. Person B will immediately overwrite this new file with an older version because he is one-hour forward and things the server has an old copy. 

This is can be resolved with a simple versioning system, ignoring the timezone or modification time at all. It has to be fixed in the server and not in the clients, because only the server sees all clients and can establish a time line.

If somebody works on an old version of a file and tries to "check it in" after someone else has modified it on the server, the client must detect this and forbid it.

SVN, GIT and other versioning systems are dealing with this all the time, so it should not be that difficult for us.

I propose this as a simple solution and I'll use a simple git/svn lingo: 

Use the server to handle all versions of a file. When a file is "upload" to the server, the version number increases (by 1). For every version number, the server will save a hash of the file contents too, so your key is Version+Hash (to verify file integrity).

When a client checks-out a file, it must know what version was checked out. During sync, the client can always tell if a file was modified locally before overwriting it by another version by using the hash and the version number.

During upload to the server, the server will easily know if the uploaded file is a next major version of if the client is uploading a modified version of an old copy of the file. The same goes for the client. It will know if everything is right or if there was a newer version on the server.

This all works nicely, until two persons (to clients) modify the same file at the same time. The first one to upload to the server will win. The second will be warned about conflicts and will be left with a choise to either upload a conflicted copy (i.e. making 2 copies on the server, one named "conflicted copy") or, instead, download the newer version from the server locally and end up with two local files to resolve the conflicts by hand.

This is the only way to deal consistently with multiple updates to the same files. You cannot simply merge binary files (like we do with git source code) and you cannot trust the server to rule who wins.

This is core functionality which should be inside the sync protocol from start (imho).

**Comment by nickvergessen** [comment]
No, that is not really a problem and also we have versioning in place. Also this doesn't happen accidentally.

**Comment by HEG-IT** [comment]
I'm not sure how we conclude that this isn't a problem as we're still seeing it in our environment and it's keeping us from utilizing the Windows Desktop client.

We're using nextcloud to provide access to internal file shares (located in timezone "0") by staff that are in timezones ["0","1","3"]. Employee in timezone "3" updates a file, which the desktop client syncs through nextcloud to the server. The server updates the timestamp to reflect the local timezone (timezone "0"), which then makes the file "newer" to nextcloud and the actual source gets clobbered in the next sync cycle. Right now the "solution" is that our timezone "1" and "3" have to pause the synce until the timezone differential has passed and then resume the sync so their files don't get corrupted. Not really a solution when we were hoping for a much tighter sync window.

I'm not opposed to the solution suggested by @pquan, but I think it might be overkill. I'd suggest that the server just deal in UTC. The server would have to perform the UTC calculation for all the files that it directly accesses, but the client would handle the calculation for the files to be synced. With the sync comparison being done in UTC the clobber problem goes away and sync can become close to synchronous like we'd expect.

This shortcoming is becoming an increasing problem for us and I'm to the point that I'd like to know what it would cost to have it fixed. Because the cost of having it not work is not trivial.

**Comment by pquan** [comment]
Perhaps a complete versioning system is overkill. Working in UTC on the server side and precisely syncing time between clients and server would suffice most of the time but would expose us to race conditions where multiple clients are not perfectly synced or have undetected clock drift.

A simpler solution could be that the server attaches time-ordered tickets on each file, and that the clients know which version of the file they want to upload or sync to back to the server.

So, each download from Server->Client will have an incremental time-based ticket attached, containing the last update time of the file *on the server*. This ticket will be conserved by the client and will be used when uploading backup. 

So, during a sync-back from Client to Server the client would use this time-based ticket. 
For the server it becomes trivial to know if a client is uploading a file that has an older ticket than the currently available on the server, i.e. the server now knows if another client uploaded a modified version before this particular client and can act accordingly.

**Comment by szaimen** [comment]
Hi, please update to at least 23.0.12 and report back if it fixes the issue. Thank you!
